subprojects/persistent-cache/src/main/java/org/gradle/cache/internal/AbstractCacheCleanup.java

```diff
@@ -0,0 +1,138 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.cache.internal;
+
+import org.apache.commons.io.FileUtils;
+import org.gradle.api.Action;
+import org.gradle.cache.PersistentCache;
+import org.gradle.internal.operations.BuildOperationContext;
+import org.gradle.internal.operations.BuildOperationExecutor;
+import org.gradle.internal.operations.CallableBuildOperation;
+import org.gradle.internal.operations.RunnableBuildOperation;
+import org.gradle.internal.progress.BuildOperationDescriptor;
+import org.gradle.util.GFileUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.FilenameFilter;
+import java.util.List;
+
+abstract class AbstractCacheCleanup implements Action<PersistentCache> {
+    private static final Logger LOGGER = LoggerFactory.getLogger(AbstractCacheCleanup.class);
+
+    private final BuildOperationExecutor buildOperationExecutor;
+
+    public AbstractCacheCleanup(BuildOperationExecutor buildOperationExecutor) {
+        this.buildOperationExecutor = buildOperationExecutor;
+    }
+
+    @Override
+    public void execute(final PersistentCache persistentCache) {
+        buildOperationExecutor.run(new RunnableBuildOperation() {
+            @Override
+            public void run(BuildOperationContext context) {
+                cleanup(persistentCache);
+            }
+
+            @Override
+            public BuildOperationDescriptor.Builder description() {
+                return BuildOperationDescriptor.displayName("Clean up " + persistentCache);
+            }
+        });
+
+    }
+
+    private void cleanup(final PersistentCache persistentCache) {
+        final File[] filesEligibleForCleanup = buildOperationExecutor.call(new CallableBuildOperation<File[]>() {
+            @Override
+            public File[] call(BuildOperationContext context) {
+                return findEligibleFiles(persistentCache.getBaseDir());
+            }
+
+            @Override
+            public BuildOperationDescriptor.Builder description() {
+                return BuildOperationDescriptor.displayName("Scan " + persistentCache.getBaseDir());
+            }
+        });
+
+        if (filesEligibleForCleanup.length > 0) {
+            final List<File> filesForDeletion = buildOperationExecutor.call(new CallableBuildOperation<List<File>>() {
+                @Override
+                public List<File> call(BuildOperationContext context) {
+                    return findFilesToDelete(persistentCache, filesEligibleForCleanup);
```

Comment:
I'd probably just merge this with the previous build operation, or even just merge them all into 1.

Reply:
Yeah, I think a single build operation would work here.